linera-core/src/worker.rs

```diff
@@ -1091,34 +1060,59 @@ where
                 recipient,
                 certificates,
             } => {
-                let mut actions = NetworkActions::default();
+                let mut latest_heights = Vec::<(Medium, BlockHeight)>::new();
                 for (medium, heights) in height_map {
-                    let origin = Origin { sender, medium };
+                    let origin = Origin {
+                        sender,
+                        medium: medium.clone(),
+                    };
                     let app_certificates = certificates
                         .iter()
                         .filter(|cert| heights.binary_search(&cert.value.block().height).is_ok())
                         .cloned()
                         .collect();
-                    actions.merge(
-                        self.process_cross_chain_update(
-                            origin.clone(),
-                            recipient,
-                            app_certificates,
-                        )
-                        .await?,
-                    );
+                    if let Some(height) = self
+                        .process_cross_chain_update(origin.clone(), recipient, app_certificates)
+                        .await?
+                    {
+                        latest_heights.push((medium, height));
+                    }
                 }
-                Ok(actions)
+                if latest_heights.is_empty() {
+                    return Ok(NetworkActions::default());
+                }
+                let notifications = latest_heights
+                    .iter()
+                    .cloned()
+                    .map(|(medium, height)| {
+                        let origin = Origin { sender, medium };
+                        Notification {
+                            chain_id: recipient,
+                            reason: Reason::NewMessage { origin, height },
+                        }
+                    })
+                    .collect();
+                let cross_chain_requests = vec![CrossChainRequest::ConfirmUpdatedRecipient {
+                    sender,
+                    recipient,
+                    latest_heights,
+                }];
+                Ok(NetworkActions {
+                    cross_chain_requests,
+                    notifications,
+                })
             }
             CrossChainRequest::ConfirmUpdatedRecipient {
-                origin: Origin { sender, medium },
+                sender,
                 recipient,
-                height,
+                latest_heights,
             } => {
                 let mut chain = self.storage.load_chain(sender).await?;
-                let target = Target { recipient, medium };
-                if chain.mark_messages_as_received(target, height).await? {
-                    if chain.all_messages_delivered_up_to(height) {
+                for (medium, height) in latest_heights {
```

Comment:
Need to move `chain.save().await` out of the main loop (with a test that there is something to write).
Removing these sequential DB accesses might very well be the main gain of this PR.